File: banki/srs.py
"""SM-2 spaced repetition algorithm.

Based on the SuperMemo SM-2 algorithm with Anki-style 4-button rating:
    Again (1) - failed, reset interval
    Hard  (2) - barely remembered
    Good  (3) - remembered with effort
    Easy  (4) - trivial recall
"""
from dataclasses import dataclass
from datetime import date, timedelta

AGAIN = 1
HARD = 2
GOOD = 3
EASY = 4

RATING_LABELS = {
    AGAIN: "다시",
    HARD: "어려움",
    GOOD: "괜찮음",
    EASY: "쉬움",
}

RATING_COLORS = {
    AGAIN: "#ef4444",
    HARD: "#f59e0b",
    GOOD: "#22c55e",
    EASY: "#3b82f6",
}

DEFAULT_EASE = 2.5
MIN_EASE = 1.3
MAX_EASE = 3.0


@dataclass
class CardState:
    ease: float = DEFAULT_EASE
    interval_days: int = 0
    repetitions: int = 0
    last_reviewed: date | None = None
    due_date: date | None = None
    lapses: int = 0

    @property
    def is_new(self) -> bool:
        return self.repetitions == 0 and self.last_reviewed is None

    def is_due(self, today: date | None = None) -> bool:
        today = today or date.today()
        if self.due_date is None:
            return True
        return self.due_date <= today


def review(state: CardState, rating: int, today: date | None = None) -> CardState:
    """Apply a rating to a card and return its new state."""
    today = today or date.today()
    new = CardState(
        ease=state.ease,
        interval_days=state.interval_days,
        repetitions=state.repetitions,
        last_reviewed=today,
        due_date=state.due_date,
        lapses=state.lapses,
    )

    if rating == AGAIN:
        new.repetitions = 0
        new.interval_days = 1
        new.lapses = state.lapses + 1
        new.ease = max(MIN_EASE, state.ease - 0.20)
    else:
        new.repetitions = state.repetitions + 1
        if new.repetitions == 1:
            new.interval_days = 1
        elif new.repetitions == 2:
            new.interval_days = 6
        else:
            multiplier = state.ease
            if rating == HARD:
                multiplier = max(MIN_EASE, state.ease * 0.8)
            elif rating == EASY:
                multiplier = state.ease * 1.3
            new.interval_days = max(1, round(state.interval_days * multiplier))

        if rating == HARD:
            new.ease = max(MIN_EASE, state.ease - 0.15)
        elif rating == GOOD:
            new.ease = state.ease
        elif rating == EASY:
            new.ease = min(MAX_EASE, state.ease + 0.15)

    new.due_date = today + timedelta(days=new.interval_days)
    return new


def rating_from_score(score: int) -> int:
    """Suggest a rating from a dictation score percentage."""
    if score >= 95:
        return EASY
    if score >= 80:
        return GOOD
    if score >= 50:
        return HARD
    return AGAIN


def sort_for_session(card_states: dict[int, CardState], all_indices: list[int], today: date | None = None) -> list[int]:
    """Return a learning order: due cards first (by overdue-ness), then new cards."""
    today = today or date.today()
    due = []
    new = []
    future = []
    for idx in all_indices:
        s = card_states.get(idx)
        if s is None or s.is_new:
            new.append(idx)
        elif s.is_due(today):
            overdue = (today - (s.due_date or today)).days
            due.append((-overdue, idx))
        else:
            future.append(idx)
    due.sort()
    return [i for _, i in due] + new + future

File: web/js/lib/srs.js
// SM-2 spaced repetition, ported from banki/srs.py.
//
// Anki-style four-button rating; card state keeps the Python field names so it
// round-trips through storage and the fixture tests unchanged. Dates are
// YYYY-MM-DD strings in the learner's local calendar.

import { pyRound, todayISO, addDays, daysBetween } from './util.js';

export const AGAIN = 1;
export const HARD = 2;
export const GOOD = 3;
export const EASY = 4;

export const RATING_LABELS = {
  [AGAIN]: '다시',
  [HARD]: '어려움',
  [GOOD]: '괜찮음',
  [EASY]: '쉬움',
};

export const DEFAULT_EASE = 2.5;
export const MIN_EASE = 1.3;
export const MAX_EASE = 3.0;

export function newCardState() {
  return {
    ease: DEFAULT_EASE,
    interval_days: 0,
    repetitions: 0,
    last_reviewed: null,
    due_date: null,
    lapses: 0,
  };
}

export function isNew(state) {
  return state.repetitions === 0 && !state.last_reviewed;
}

export function isDue(state, today = todayISO()) {
  if (!state.due_date) return true;
  return state.due_date <= today;
}

/** Apply a rating to a card and return its new state. */
export function review(state, rating, today = todayISO()) {
  const next = { ...state, last_reviewed: today };

  if (rating === AGAIN) {
    next.repetitions = 0;
    next.interval_days = 1;
    next.lapses = state.lapses + 1;
    next.ease = Math.max(MIN_EASE, state.ease - 0.2);
  } else {
    next.repetitions = state.repetitions + 1;
    if (next.repetitions === 1) {
      next.interval_days = 1;
    } else if (next.repetitions === 2) {
      next.interval_days = 6;
    } else {
      let multiplier = state.ease;
      if (rating === HARD) multiplier = Math.max(MIN_EASE, state.ease * 0.8);
      else if (rating === EASY) multiplier = state.ease * 1.3;
      next.interval_days = Math.max(1, pyRound(state.interval_days * multiplier));
    }

    if (rating === HARD) next.ease = Math.max(MIN_EASE, state.ease - 0.15);
    else if (rating === GOOD) next.ease = state.ease;
    else if (rating === EASY) next.ease = Math.min(MAX_EASE, state.ease + 0.15);
  }

  next.due_date = addDays(today, next.interval_days);
  return next;
}

/** Suggest a rating from a dictation score percentage. */
export function ratingFromScore(score) {
  if (score >= 95) return EASY;
  if (score >= 80) return GOOD;
  if (score >= 50) return HARD;
  return AGAIN;
}

/** Learning order: due cards first (most overdue first), then new, then future. */
export function sortForSession(cardStates, allIndices, today = todayISO()) {
  const due = [];
  const fresh = [];
  const future = [];

  allIndices.forEach((idx) => {
    const state = cardStates[idx];
    if (!state || isNew(state)) {
      fresh.push(idx);
    } else if (isDue(state, today)) {
      const overdue = daysBetween(state.due_date || today, today);
      due.push([-overdue, idx]);
    } else {
      future.push(idx);
    }
  });

  due.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
  return [...due.map(([, idx]) => idx), ...fresh, ...future];
}
